ambry-replication/src/main/java/com/github/ambry/replication/ReplicationManager.java

```diff
@@ -386,33 +451,45 @@ public void onPartitionBecomeStandbyFromBootstrap(String partitionName) {
+        // 1. get replica ID of current node from store manager
+        ReplicaId localReplica = storeManager.getReplica(partitionName);
+
+        // 2. Get the peer leader replicas from all data centers for this partition
+        List<? extends ReplicaId> leaderReplicas =
+            localReplica.getPartitionId().getReplicaIdsByState(ReplicaState.LEADER, null);
 
-      // 2. Get the peer leader replicas from all data centers for this partition
-      List<? extends ReplicaId> leaderReplicas =
-          localReplica.getPartitionId().getReplicaIdsByState(ReplicaState.LEADER, null);
-
-      // 3. Log the list of leader replicas associated with this partition (will be used later for leadership based replication)
-      List<ReplicaId> peerLeaderReplicas = new ArrayList<>();
-      for (ReplicaId leaderReplica : leaderReplicas) {
-        if (leaderReplica.getDataNodeId() != localReplica.getDataNodeId()) {
-          peerLeaderReplicas.add(leaderReplica);
-          logger.info("Partition {} on node instance {} is leader in remote dc {}", partitionName,
-              getInstanceName(leaderReplica.getDataNodeId().getHostname(), leaderReplica.getDataNodeId().getPort()),
-              leaderReplica.getDataNodeId().getDatacenterName());
+        // 3. Log the list of leader replicas associated with this partition (will be used later for leadership based replication)
+        List<ReplicaId> peerLeaderReplicas = new ArrayList<>();
+        for (ReplicaId leaderReplica : leaderReplicas) {
+          if (leaderReplica.getDataNodeId() != localReplica.getDataNodeId()) {
+            peerLeaderReplicas.add(leaderReplica);
+            logger.info("Partition {} on node instance {} is leader in remote dc {}", partitionName,
+                getInstanceName(leaderReplica.getDataNodeId().getHostname(), leaderReplica.getDataNodeId().getPort()),
+                leaderReplica.getDataNodeId().getDatacenterName());
+          }
         }
+        peerLeaderReplicasByPartition.put(partitionName, peerLeaderReplicas);
+      } finally {
+        rwlock1.writeLock().unlock();
       }
-      peerLeaderReplicasByPartition.put(partitionName, peerLeaderReplicas);
     }
 
     @Override
     public void onPartitionBecomeStandbyFromLeader(String partitionName) {
       logger.info("Partition state change notification from Leader to Standby received for partition {}",
           partitionName);
-      if (peerLeaderReplicasByPartition.containsKey(partitionName)) {
-        peerLeaderReplicasByPartition.remove((partitionName));
+      // Read-write lock to avoid contention between this thread (where partition are removed from the map) and thread calling onRoutingTableUpdate() (where partitions are read and updated to map).
+      rwlock1.writeLock().lock();
+      try {
+        if (peerLeaderReplicasByPartition.containsKey(partitionName)) {
+          peerLeaderReplicasByPartition.remove((partitionName));
```

Comment:
why we need to check if `peerLeaderReplicasByPartition` contains `partitionName`? I feel  it's unnecessary.